### PR TITLE
feature: supported OpenSSL 1.1.0

### PR DESCRIPTION
**Pull request by spacewander**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

For #903 and #1172.

Changes contains:
1. Take changes from #761 and #922.
2. Update the error message.
3. Compared to 1.0.2, OpenSSL 1.1.0 SSL_CTX_set_verify_depth verifies
one more level. Since Nginx doesn't change the default value for
ssl_verfiy_depth, I just update the test cases instead of updating the
C code.
4. Add OpenSSL 1.1.0c to the travis build.

### CODE REVIEW

**Comment by spacewander** [comment]
I will send a pr to stream-lua-nginx-module if this one is able to merge.

File: t/129-ssl-socket.t [hunk 1]
```diff
@@ -693,7 +693,7 @@ SSL reused session
     server_tokens off;
     resolver $TEST_NGINX_RESOLVER ipv6=off;
     lua_ssl_trusted_certificate ../html/trusted.crt;
-    lua_ssl_verify_depth 1;
+    lua_ssl_verify_depth 0;
```
Comment (agentzh) [inline t/129-ssl-socket.t]: Hmm, why is this necessary for OpenSSL 1.1?

File: t/129-ssl-socket.t [hunk 2]
```diff
@@ -771,7 +772,7 @@ SSL reused session
...
```
Comment (agentzh) [inline t/129-ssl-socket.t]: Ditto.

File: t/129-ssl-socket.t (hunk 1 above)
Comment (spacewander) [reply t/129-ssl-socket.t]: @agentzh 
`SSL_CTX_set_verify_depth` in OpenSSL 1.1.0 verifies
one more level. Since Nginx doesn't change the default value for
ssl_verfiy_depth, I just update the test cases instead of updating the
C code.

Comment (agentzh) [reply t/129-ssl-socket.t]: @spacewander I see. Thanks for your explanation!

**Comment by agentzh** [comment]
Merged. Thanks! I have to (temporarily) comment out the openssl 1.1.0c test case in the travis ci config file since `ngx_stream_lua` does not work with this openssl version series yet.

**Comment by neilstuartcraig** [comment]
This is awesome news! I'll be super excited to see this work in a release :-)
Thanks so much @spacewander, you are an absolute legend for doing this - i am sure it wasn't simple or quick! Thanks also to everyone else who helped with it.

File: src/ngx_http_lua_ssl_session_storeby.c [hunk 3]
```diff
@@ -247,11 +249,13 @@ ngx_http_lua_ssl_sess_store_handler(ngx_ssl_conn_t *ssl_conn,
         }
     }
 
+    sess_id = SSL_SESSION_get_id(sess, &sess_id_len);
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_session_storeby.c]: This change breaks compatibility with older OpenSSL like 0.9.x.

Comment (agentzh) [reply src/ngx_http_lua_ssl_session_storeby.c]: Just fixed in commit 7c8c503001.

**Comment by S4mw1s3** [comment]
A big thank you to all of the people who helped implementing support for OpenSSL 1.1.0 !

**Comment by thibaultcha** [comment]
Great job @ghedo @doujiang24 @spacewander, thank you! 👍

**Comment by frankgould** [comment]
I am trying to upgrade my RPi mate 16.04 with openssl 1.1.0f but after doing so, I cannot log in because I get 'failed to start session' error. Is there something I did wrong in the install and a different way to prevent this?
